File: ios/Vocal/UI/AudioStepTimeline.swift
import SwiftUI

struct AudioStepTimeline: View {
  let steps: [AudioStepPresentation]
  var timings: ServerStageTimings?

  var body: some View {
    VStack(alignment: .leading, spacing: 0) {
      ForEach(Array(steps.enumerated()), id: \.element.id) { index, step in
        HStack(alignment: .top, spacing: 12) {
          VStack(spacing: 0) {
            Image(systemName: symbol(step.state))
              .foregroundStyle(color(step.state)).frame(width: 24, height: 24)
            if index < steps.count - 1 {
              Rectangle().fill(color(step.state).opacity(0.35)).frame(width: 2, height: 30)
            }
          }
          VStack(alignment: .leading, spacing: 3) {
            Text(LocalizedStringKey(step.titleKey)).font(.subheadline.weight(.semibold))
            if let date = step.occurredAt {
              Text(date, style: .time).font(.caption).foregroundStyle(.secondary)
            }
          }.padding(.top, 2)
          Spacer()
          if let measurement = step.measurement, measurement.durationMs >= 0 {
            let duration = String(format: "%.1f s", Double(measurement.durationMs) / 1_000)
            Text(
              measurement.complete
                ? duration : String(format: String(localized: "timing_partial"), duration)
            )
            .font(.caption.monospacedDigit()).foregroundStyle(.secondary)
          }
        }
        .accessibilityElement(children: .combine)
      }
      if let timings {
        Divider().padding(.vertical, 12)
        Text("server_timings").font(.headline)
        Text("timing_note").font(.caption).foregroundStyle(.secondary).padding(.bottom, 8)
        ForEach(Array(timings.stages.enumerated()), id: \.offset) { _, measurement in
          if measurement.durationMs >= 0 {
            timingRow(stageTitle(measurement.stage), measurement.durationMs, measurement.complete)
          }
        }
        if let total = timings.totalMs, total >= 0 {
          timingRow("server_total", total, timings.totalComplete)
        }
      }
    }
  }

  private func timingRow(_ title: String, _ milliseconds: Int64, _ complete: Bool) -> some View {
    HStack(alignment: .firstTextBaseline) {
      Text(LocalizedStringKey(title))
      Spacer()
      let duration = String(format: "%.1f s", Double(milliseconds) / 1_000)
      Text(complete ? duration : String(format: String(localized: "timing_partial"), duration))
        .monospacedDigit()
    }.font(.subheadline).padding(.vertical, 4)
  }

  private func stageTitle(_ stage: String) -> String {
    let known = [
      "import-queue", "source-download", "source-validation", "source-upload",
      "upload-confirmation", "submission-window", "queue", "retry-wait", "resource-check",
      "input-download", "input-validation", "preparation", "model-load", "separation", "denoise",
      "trim", "encoding", "output-validation", "output-ready", "output-upload", "completion",
    ]
    return known.contains(stage)
      ? "timing_stage_" + stage.replacingOccurrences(of: "-", with: "_") : "timing_stage_other"
  }

  private func symbol(_ state: AudioStepState) -> String {
    switch state {
    case .complete: return "checkmark.circle.fill"
    case .active: return "waveform.circle.fill"
    case .failed: return "exclamationmark.circle.fill"
    case .cancelled: return "xmark.circle.fill"
    case .pending: return "circle"
    }
  }
  private func color(_ state: AudioStepState) -> Color {
    switch state {
    case .failed: return .red
    case .cancelled: return .secondary
    case .pending: return .secondary.opacity(0.55)
    default: return VocalStyle.teal
    }
  }
}

File: android/app/src/main/java/com/hatem/musicmute/ui/AudioStepTimeline.kt
package com.hatem.musicmute.ui

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Cancel
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material.icons.outlined.PauseCircleOutline
import androidx.compose.material.icons.outlined.RadioButtonChecked
import androidx.compose.material.icons.outlined.RadioButtonUnchecked
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.foundation.background
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.hatem.musicmute.R
import com.hatem.musicmute.processing.AudioTaskPresentation
import com.hatem.musicmute.processing.AudioStepState
import com.hatem.musicmute.processing.audioTaskTimeline
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import com.hatem.musicmute.ui.design.CreativeTokens

@Composable
fun AudioStepTimeline(task: AudioTaskPresentation) {
    val steps = audioTaskTimeline(task)
    Column(verticalArrangement = Arrangement.spacedBy(CreativeTokens.CompactGap)) {
        Text(stringResource(R.string.audio_task_timeline), style = MaterialTheme.typography.titleMedium)
        Column {
            steps.forEachIndexed { index, step ->
                val status = stringResource(when (step.state) {
                    AudioStepState.COMPLETE -> R.string.ui_step_complete
                    AudioStepState.CURRENT -> R.string.ui_step_current
                    AudioStepState.PENDING -> R.string.ui_step_pending
                    else -> audioTaskStageLabel(task.stage)
                })
                val color = when (step.state) {
                    AudioStepState.COMPLETE, AudioStepState.CURRENT -> MaterialTheme.colorScheme.primary
                    AudioStepState.FAILED -> MaterialTheme.colorScheme.error
                    else -> MaterialTheme.colorScheme.outline
                }
                Row(
                    Modifier.fillMaxWidth().height(IntrinsicSize.Min)
                        .semantics(mergeDescendants = true) { stateDescription = status },
                    verticalAlignment = Alignment.Top,
                    horizontalArrangement = Arrangement.spacedBy(CreativeTokens.CompactGap),
                ) {
                    Column(Modifier.width(24.dp).fillMaxHeight(), horizontalAlignment = Alignment.CenterHorizontally) {
                        Icon(
                            when (step.state) {
                                AudioStepState.COMPLETE -> Icons.Outlined.CheckCircle
                                AudioStepState.CURRENT -> Icons.Outlined.RadioButtonChecked
                                AudioStepState.INTERRUPTED -> Icons.Outlined.PauseCircleOutline
                                AudioStepState.FAILED -> Icons.Outlined.ErrorOutline
                                AudioStepState.CANCELLED -> Icons.Outlined.Cancel
                                AudioStepState.PENDING -> Icons.Outlined.RadioButtonUnchecked
                            },
                            contentDescription = null,
                            modifier = Modifier.size(24.dp),
                            tint = color,
                        )
                        if (index != steps.lastIndex) Box(
                            Modifier.width(1.dp).weight(1f).heightIn(min = 16.dp).background(
                                if (step.state == AudioStepState.COMPLETE) MaterialTheme.colorScheme.primary
                                else MaterialTheme.colorScheme.outlineVariant),
                        )
                    }
                    Text(
                        stringResource(audioTaskStageLabel(step.stage)),
                        modifier = Modifier.weight(1f).padding(bottom = if (index == steps.lastIndex) 0.dp else 16.dp),
                        style = if (step.state == AudioStepState.CURRENT) MaterialTheme.typography.titleSmall
                            else MaterialTheme.typography.bodyMedium,
                        color = if (step.state == AudioStepState.PENDING) MaterialTheme.colorScheme.onSurfaceVariant
                            else MaterialTheme.colorScheme.onSurface,
                    )
                    step.measurement?.takeIf { it.durationMs >= 0 }?.let {
                        val duration = String.format(java.util.Locale.getDefault(), "%.1f s", it.durationMs / 1000.0)
                        Text(if (it.complete) duration else stringResource(R.string.timing_partial, duration),
                            style = MaterialTheme.typography.labelSmall)
                    }
                }
            }
        }
        task.serverStageTimings?.let { timing ->
            HorizontalDivider()
            Text(stringResource(R.string.server_timings), style = MaterialTheme.typography.titleMedium)
            Text(stringResource(R.string.timing_note), style = MaterialTheme.typography.bodySmall)
            timing.stages.filter { it.durationMs >= 0 }.forEach { measurement ->
                ServerTimingRow(stringResource(serverTimingLabel(measurement.stage)), measurement.durationMs, measurement.complete)
            }
            timing.totalMs?.takeIf { it >= 0 }?.let {
                ServerTimingRow(stringResource(R.string.server_total), it, timing.totalComplete)
            }
        }
    }
}

@Composable
private fun ServerTimingRow(label: String, durationMs: Long, complete: Boolean) {
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(label, Modifier.weight(1f), style = MaterialTheme.typography.bodyMedium)
        val duration = String.format(java.util.Locale.getDefault(), "%.1f s", durationMs / 1000.0)
        Text(if (complete) duration else stringResource(R.string.timing_partial, duration),
            style = MaterialTheme.typography.bodyMedium.copy(fontFamily = androidx.compose.ui.text.font.FontFamily.Monospace))
    }
}

private fun serverTimingLabel(stage: String): Int = when (stage) {
    "import-queue" -> R.string.timing_stage_import_queue
    "source-download" -> R.string.timing_stage_source_download
    "source-validation" -> R.string.timing_stage_source_validation
    "source-upload" -> R.string.timing_stage_source_upload
    "upload-confirmation" -> R.string.timing_stage_upload_confirmation
    "submission-window" -> R.string.timing_stage_submission_window
    "queue" -> R.string.timing_stage_queue
    "retry-wait" -> R.string.timing_stage_retry_wait
    "resource-check" -> R.string.timing_stage_resource_check
    "input-download" -> R.string.timing_stage_input_download
    "input-validation" -> R.string.timing_stage_input_validation
    "preparation" -> R.string.timing_stage_preparation
    "model-load" -> R.string.timing_stage_model_load
    "separation" -> R.string.timing_stage_separation
    "denoise" -> R.string.timing_stage_denoise
    "trim" -> R.string.timing_stage_trim
    "encoding" -> R.string.timing_stage_encoding
    "output-validation" -> R.string.timing_stage_output_validation
    "output-ready" -> R.string.timing_stage_output_ready
    "output-upload" -> R.string.timing_stage_output_upload
    "completion" -> R.string.timing_stage_completion
    else -> R.string.timing_stage_other
}
